modules/tensor_mechanics/include/materials/FiniteStrainUObasedCP.h

```diff
@@ -0,0 +1,295 @@
+ * Backward Euler integration rule is used for the rate equations.
+ *
+ * Involves 4 different types of user objects that calculates:
+ * State variables - update state variable (derive from CrystalPlasticityStateVariable)
+ * State variable evolution compoment - individual component of state variable incremental rate (derive from CrystalPlasticityStateVariableEvolutionRateComponent)
+ * Slip resistance - calcuate slip resistances (derive from CrystalPlasticitySlipResistances)
+ * Slip rates - calcuate flow direction and slip rates (derive from CrystalPlasticitySlipRates)
+ */
+class FiniteStrainUObasedCP : public ComputeStressBase
+{
+public:
+  FiniteStrainUObasedCP(const InputParameters & parameters);
+
+protected:
+  /**
+   * updates the stress at a quadrature point.
+   */
+  virtual void computeQpStress();
+
+  /**
+   * updates the elasticity tensor at a quadrature point.
+   * Presently void.
+   */
+  virtual void computeQpElasticityTensor();
+
+  /**
+   * initializes the stateful properties such as
+   * stress, plastic deformation gradient, slip system resistances, etc.
+   */
+  virtual void initQpStatefulProperties();
+
+  /**
+   * calls the residual and jacobian functions used in the
+   * stress update algorithm.
+   */
+  virtual void calcResidJacob(RankTwoTensor &, RankFourTensor &);
+
+  /**
+   * updates the slip system resistances.
+   * Override to modify slip system resistance evolution
+   */
+  virtual void updateSlipSystemResistance();
+
+  /**
+   * set variables for stress and internal variable solve.
+   */
+  virtual void preSolveQp();
+
+  /**
+   * solve stress and internal variables.
+   */
+  virtual void solveQp();
+
+  /**
+   * update stress and internal variable after solve.
+   */
+  virtual void postSolveQp();
+
+  /**
+   * set variables for internal variable solve.
+   */
+  virtual void preSolveStatevar();
+
+  /**
+   * solve internal variables.
+   */
+  virtual void solveStatevar();
+
+  /**
+   * update internal variable after solve.
+   */
+  virtual void postSolveStatevar();
+
+  /**
+   * set variables for stress solve.
+   */
+  virtual void preSolveStress();
+
+  /**
+   * solves for stress, updates plastic deformation gradient.
+   */
+  virtual void solveStress();
+
+  /**
+   * update stress and plastic deformation gradient after solve.
+   */
+  virtual void postSolveStress();
+
+  /**
+   * calculate stress residual.
+   */
+  virtual void calcResidual( RankTwoTensor & );
+
+  /**
+   * calculate jacobian.
+   */
+  virtual void calcJacobian( RankFourTensor & );
+
+  /**
+   * updates the slip increments.
+  */
+  virtual void getSlipIncrements();
+
+  /**
+   * calculate the tangent moduli for preconditioner.
+   * Default is the elastic stiffness matrix.
+   * Exact jacobian is currently implemented.
+   * tan_mod_type can be modified to exact in .i file to turn it on.
+   */
+  virtual ElasticityTensorR4 calcTangentModuli();
```

Comment:
Can make the TangentModuli calculating functions return void since member variables are used anyways.